drivers/flash/spi_nor.c

```diff
@@ -0,0 +1,453 @@
+#define SYNC_LOCK() k_sem_take(&driver_data->sem, K_FOREVER)
+#define SYNC_UNLOCK() k_sem_give(&driver_data->sem)
+#else
+#define SYNC_INIT()
+#define SYNC_LOCK()
+#define SYNC_UNLOCK()
+#endif
+
+/*
+ * @brief Send an SPI command
+ *
+ * @param dev Device struct
+ * @param opcode The command to send
+ * @param is_addressed A flag to define if the command is addressed
+ * @param addr The address to send
+ * @param data The buffer to store or read the value
+ * @param length The size of the buffer
+ * @param is_write A flag to define if it's a read or a write command
+ * @return 0 on success, negative errno code otherwise
+ */
+static int spi_nor_access(const struct device *const dev,
+			  u8_t opcode, bool is_addressed, off_t addr,
+			  void *data, size_t length, bool is_write)
+{
+	struct spi_nor_data *const driver_data = dev->driver_data;
+
+	u8_t buf[4] = {
+		opcode,
+		(addr & 0xFF0000) >> 16,
+		(addr & 0xFF00) >> 8,
+		(addr & 0xFF),
+	};
+
+	struct spi_buf spi_buf[2] = {
+		{
+			.buf = buf,
+			.len = (is_addressed) ? 4 : 1,
+		},
+		{
+			.buf = data,
+			.len = length
+		}
+	};
+	const struct spi_buf_set tx_set = {
+		.buffers = spi_buf,
+		.count = (length) ? 2 : 1
+	};
+
+	const struct spi_buf_set rx_set = {
+		.buffers = spi_buf,
+		.count = 2
+	};
+
+	if (is_write) {
+		return spi_write(driver_data->spi,
+			&driver_data->spi_cfg, &tx_set);
+	}
+
+	return spi_transceive(driver_data->spi,
+		&driver_data->spi_cfg, &tx_set, &rx_set);
+}
+#define spi_nor_cmd_read(dev, opcode, dest, length) \
+	spi_nor_access(dev, opcode, false, 0, dest, length, false)
+#define spi_nor_cmd_addr_read(dev, opcode, addr, dest, length) \
+	spi_nor_access(dev, opcode, true, addr, dest, length, false)
+#define spi_nor_cmd_write(dev, opcode) \
+	spi_nor_access(dev, opcode, false, 0, NULL, 0, true)
+#define spi_nor_cmd_addr_write(dev, opcode, addr, src, length) \
+	spi_nor_access(dev, opcode, true, addr, src, length, true)
+
+/**
+ * @brief Retrieve the Flash JEDEC ID and compare it with the one expected
+ *
+ * @param dev The device structure
+ * @param flash_id The flash info structure which contains the expected JEDEC ID
+ * @return 0 on success, negative errno code otherwise
+ */
+static inline int spi_nor_read_id(struct device *dev,
+				  const struct spi_nor_config *const flash_id)
+{
+	u8_t buf[SPI_NOR_MAX_ID_LEN];
+
+	if (spi_nor_cmd_read(dev, SPI_NOR_CMD_RDID, buf,
+	    SPI_NOR_MAX_ID_LEN) != 0) {
+		return -EIO;
+	}
+
+	if (memcmp(flash_id->id, buf, SPI_NOR_MAX_ID_LEN) != 0) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Wait until the flash is ready
+ *
+ * @param dev The device structure
+ * @return 0 on success, negative errno code otherwise
+ */
+static inline int spi_nor_wait_until_ready(struct device *dev)
+{
+	int ret;
+	u8_t reg;
+
+	do {
+		ret = spi_nor_cmd_read(dev, SPI_NOR_CMD_RDSR, &reg, 1);
+	} while (!ret && (reg & SPI_NOR_WIP_BIT));
+
+	return ret;
+}
+
+/**
```

Comment:
remove all flash API documentation: it's redundant with the generic flash header.